lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp

```diff
@@ -461,6 +473,74 @@ CFEPAnalyzer<FunctionOracle>::CFEPAnalyzer(llvm::Module &M,
   }
 }
 
+static FunctionSummary
+initializeSummary(llvm::Module &M,
+                  const std::vector<GlobalVariable *> &ABIRegs,
+                  const model::Function &Function) {
+  using namespace llvm;
+  using namespace model;
+  using RegisterState = model::RegisterState::Values;
+
+  FunctionSummary Summary(Function.Type,
+                          { ABIRegs.begin(), ABIRegs.end() },
+                          ABIAnalyses::ABIAnalysesResults(),
+                          {},
+                          0,
+                          nullptr);
+
+  const auto *Prototype = cast<RawFunctionType>(Function.Prototype.get());
+
+  for (auto *CSV : ABIRegs) {
+    Summary.ABIResults.ArgumentsRegisters[CSV] = RegisterState::No;
+    Summary.ABIResults.FinalReturnValuesRegisters[CSV] = RegisterState::No;
+  }
+
+  for (const auto &NTR : Prototype->Arguments) {
+    auto Name = ABIRegister::toCSVName(NTR.Location);
+    auto *CSV = cast<GlobalVariable>(M.getGlobalVariable(Name, true));
+    Summary.ABIResults.ArgumentsRegisters[CSV] = RegisterState::Yes;
+  }
+
+  for (const auto &TR : Prototype->ReturnValues) {
+    auto Name = ABIRegister::toCSVName(TR.Location);
+    auto *CSV = cast<GlobalVariable>(M.getGlobalVariable(Name, true));
+    Summary.ABIResults.FinalReturnValuesRegisters[CSV] = RegisterState::Yes;
+  }
+
+  std::set<llvm::GlobalVariable *> PreservedRegisters;
+  for (const auto &Location : Prototype->PreservedRegisters) {
+    auto Name = ABIRegister::toCSVName(Location);
+    auto *CSV = cast<GlobalVariable>(M.getGlobalVariable(Name, true));
+    PreservedRegisters.insert(CSV);
+  }
+
+  std::erase_if(Summary.ClobberedRegisters, [&](const auto &E) {
+    auto End = PreservedRegisters.end();
+    return PreservedRegisters.find(E) != End;
+  });
+
+  Summary.ElectedFSO = Prototype->FinalStackOffset;
+  return Summary;
+}
+
+using FAR = FunctionAnalysisResults;
+void FAR::initializeCache(llvm::Module &M,
+                          const std::vector<GlobalVariable *> &ABIRegs,
+                          model::Binary &Binary) {
+  using namespace llvm;
+
+  for (const model::Function &Function : Binary.Functions) {
+    if (Function.Type == model::FunctionType::Invalid)
+      continue;
+
+    // TODO: handle CABIFunctionType
+    if (!isa<model::RawFunctionType>(Function.Prototype.get()))
+      continue;
```

Comment:
Mmmh, what happens if we don't insert it?
Will we assume later on a default summary?

Reply:
Nope, no assumption on a default summary for that specific `Entry`; if it is not inserted the corresponding `Summary` will be entry.

Reply:
Mmmh, so we assume no arguments, no return values, no clobbered registers?
Are you sure we won't get a default summary at a certain point?

Reply:
Ping?

Reply:
Preferred to double-check this. We won't get a default summary if nothing is imported from the model, as it happens to be the case here. However, the function exists in the model, so an EarlyFunctionAnalysis will be scheduled, therefore, after the analysis has run, such functions (not previously imported) will have their own `Summary`.

Reply:
Why do you say that we'll analyze this function? If it has a non-`RawFunctionType` prototype (i.e., `CABIFunctionType`) but its type is not `Invalid`, we will not analyze it and we'll use the default summary for it, isn't it?

I'm not saying we need to change this, I'm trying to understand what the code does.

Reply:
I stand corrected, if the function's prototype is non-`RawFunctionType` and the type is not `Invalid`, the function will not be analyzed. If the `Oracle` is questioned on this function – for which no summary has been cached –, then the [default summary](https://github.com/revng/revng/blob/develop/lib/EarlyFunctionAnalysis/EarlyFunctionAnalysis.cpp#L268) is used.

Reply:
Closing this, as `CABIFunctionType`'s prototype is now supported.